Add seek support for streams

SlowLogicBoy opened the pull request:
Fixes #161 
TODO:
- [x] Remove no longer used code.
- [x] Fix Failing tests
- [x] Cache Stream

~~Yay! I removed more then I added 😆.~~ 😿

SlowLogicBoy commented on YoutubeExplode/Internal/YoutubeVideoStream.cs:
```diff
@@ -0,0 +1,91 @@
+        public override int Read(byte[] buffer, int offset, int count) =>
+            ReadAsync(buffer, offset, count).GetAwaiter().GetResult();
+
+        public override long Seek(long offset, SeekOrigin origin)
```
I'm not sure how Seeks work, I probably should throw instead of math'ing out what I can works with since user might get what's not expected

SlowLogicBoy replied:
[MemoryStream.Seek Implementation](https://referencesource.microsoft.com/#mscorlib/system/io/memorystream.cs,30f00da49c798fe1,references)

SlowLogicBoy commented on YoutubeExplode/Internal/YoutubeVideoStream.cs:
```diff
@@ -0,0 +1,91 @@
+
+            count = Math.Min(MaxSegmentSize - 1, count);
+            if(_currentStream == null)
+                _currentStream = await _httpClient.GetStreamAsync(_url, Position, Position + MaxSegmentSize).ConfigureAwait(false);
```
I wonder could it be that `_url` might expire before everything is downloaded?

SlowLogicBoy replied:
I also noticed that before you would call `MaxSegmentSize - 1` I'm calling it here without `-1` and it seems to work ok, should I also add it or leave it as is right now?

Tyrrrz replied:
1) No, leave it as is
2) 0 -> SegmentSize - 1; SegmentSize -> SegmentSize+SegmentSize - 1; etc..
This was done to split into even portions. In your case it doesn't matter since you're not splitting into anything beforehand. That said, technically your segments are 1 byte bigger than MaxSegmentSize so I would still add -1.

SlowLogicBoy replied:
Added -1.

Tyrrrz commented:
Oh I need a clear head to read through this. I'll try to get to this on the weekends. :)

Tyrrrz commented on YoutubeExplode/Internal/YoutubeVideoStream.cs:
```diff
@@ -0,0 +1,115 @@
+
+            var bytesRead = await _currentStream.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+            Position += bytesRead;
+            if (bytesRead == 0)
```
When is this condition met?

SlowLogicBoy replied:
When current Stream from Youtube ends, just like with [AsyncMultiStream](https://github.com/Tyrrrz/YoutubeExplode/pull/162/files#diff-0d3b577d11ab6846312caf4545e1ebcbL41)

Tyrrrz commented on YoutubeExplode/Internal/YoutubeVideoStream.cs:
```diff
@@ -0,0 +1,115 @@
+
+        private long _position;
+
+        public YoutubeVideoStream(HttpClient httpClient, string url, long length)
```
It's fine to pass `MediaStreamInfo` instead of url/length

Tyrrrz commented on YoutubeExplode/YoutubeClient.MediaStream.cs:
```diff
@@ -12,49 +12,10 @@ namespace YoutubeExplode
-                var stream = await _httpClient.GetStreamAsync(info.Url, 0, info.Size).ConfigureAwait(false);
-                return new MediaStream(info, stream);
-            }
+            return Task.FromResult(new MediaStream(info, new YoutubeVideoStream(_httpClient, info.Url, info.Size)));
```
I still think it would be nice if this was only used for rate-limited streams because otherwise it's a huge unnecessary overhead

SlowLogicBoy replied:
If I leave non rate-limited stream as it was, we won't be able to seek it, and I thought that additional overhead is not that big (additional class and couple if statements, which some of them I removed here), it will still call `_httpClient` once, as it did before (unless you seek it), the only difference (technically) will be that instead of calling `await _httpClient.GetStreamAsync(info.Url, 0, info.Size)` it will call `await _httpClient.GetStreamAsync(info.Url, 0, MaxSegmentSize)`. What do you think?

Tyrrrz replied:
I'm not sure I understand what do you mean by "it will still call `_httpClient` once". If the (non-throttled) stream has the length of 10 * MaxSegmentSize, won't that require 10 GET requests? I guess a simple solution would be to make segment size a constructor parameter and set it to `int.MaxValue` for non-throttled streams.

SlowLogicBoy replied:
Oh, I wasn't aware that non-throttled connection can have size bigger then MaxSegmentSize. I changed my code accordingly.

Tyrrrz commented:
Added some comments

Tyrrrz commented on YoutubeExplode/Internal/YoutubeVideoStream.cs:
```diff
@@ -0,0 +1,115 @@
+
+namespace YoutubeExplode.Internal
+{
+    internal class YoutubeVideoStream : Stream
```
Since it's not directly tied to Youtube in any way, can we call it `SegmentedHttpStream` or something?
It would also be nice if there was an extension method that would let me do `httpClient.GetSegmentedAsync(url)`.

SlowLogicBoy replied:
Renamed, and added an extensions.

Tyrrrz commented on YoutubeExplode/Internal/YoutubeVideoStream.cs:
```diff
@@ -0,0 +1,115 @@
+        {
+            _url = url;
+            _httpClient = httpClient;
+            Length = length;
```
Is `Length` required? Can we avoid having to pass it so that it's more versatile?

SlowLogicBoy replied:
Unfortunately we need that Length, otherwise I don't know how much youtube can give me, if I go over the boundaries `await _httpClient.GetStreamAsync(_url, Length, Length + MaxSegmentSize)` (basically when Position > Length) youtube returns 400 invalid range requested (unless you want me to catch that exception instead and say that stream is done).

I could try calculating if youtube doesn't return stream of size `MaxSegmentSize` as I request, I could say that this is the last stream they will give, but if MaxSegmentSize is exactly the size of a video, or if youtube decides to give me smaller stream then I requested anyway, it would just make it more brittle in my opinion.

SlowLogicBoy replied:
And unfortunately response headers do not return overall size, only `Content-Length` which is for current request which in our case it's `MaxSegmentSize` most of the time for us.

Tyrrrz replied:
Fair enough

Tyrrrz commented on YoutubeExplode/Internal/SegmentedHttpStream.cs:
```diff
@@ -0,0 +1,115 @@
+        public override long Position
+        {
+            get => _position;
+            set
```
What will happen when a user sets `Position`? I'm guessing since the stream isn't re-queried, it's not going to work.

SlowLogicBoy replied:
You are completely right, my bad. Fixed.

SlowLogicBoy commented:
> Hmm for some reason download time increased for
> `yield return new TestCaseData("H1O_-JVbl_k"); // very large video`
> in `Data.GetVideoIds` by significant amount after I added non rate limiting check.

My bad. Fixed.

Tyrrrz commented on YoutubeExplode/Internal/SegmentedHttpStream.cs:
```diff
@@ -0,0 +1,125 @@
+        private readonly HttpClient _httpClient;
+
+        // Maximum segment size for rate limited streams
+        private const int MaxSegmentSize = 9_898_989; // this number was carefully devised through research
```
I would prefer if this was not a constant but rather passed as a value to constructor for throttled streams

SlowLogicBoy replied:
👍

Tyrrrz commented on YoutubeExplode/YoutubeClient.MediaStream.cs:
```diff
@@ -12,49 +12,19 @@ namespace YoutubeExplode
-            // Maximum segment stream size
-            const long segmentSize = 9_898_989; // this number was carefully devised through research
-
+        private Stream GetSegmentedStreamForMediaInfo(MediaStreamInfo info)
```
No need to make a separate method for this, just inlining is fine.
If you don't want to do `Task.FromResult` you can put `await Task.Yield().ConfigureAwait(false)` before it.

SlowLogicBoy replied:
I separated it, because I was doing:
```csharp
Stream stream;
if(isRateLimited)
    stream = ...;
else
    stream = ...;
```
and I don't like unassigned variables.
of course now I have code repetition which I also don't like :)

`await Task..` adds more overhead, I prefer doing `Task.FromResult()` where there is no async operations happening, instead of forcing `async` and `await`.

Tyrrrz commented:
Awesome! Final comments.

Tyrrrz commented:
Cool, tell me when this is ready for merge.

SlowLogicBoy commented:
If you looked at my latest `Code Cleanup` commit, then you can merge after appveyor completes it's build. ;)
